### DISCUSSION TITLE
late without modulo?

### DISCUSSION THREAD

**Discussion by felixroos** [tidalcycles/strudel#133] [General]
I noticed late applies modulo 1 to its argument, which was unexpected for me.

see https://tinyurl.id/sxdbF

it would be good to have a function that doesn't do that modulo and will just delay the event by the given number of cycles, if that's even possible.

**Comment by yaxu** [comment]
Unless there's a bug, there isn't a modulo. If you have a pattern where every cycle is the same, then it follows that `late(-0.02)` will be the same as `late(0.98)`.

**Comment by yaxu** [comment]
Maybe the confusion is that you're manipulating time while also travelling through it?

**Comment by felixroos** [comment]
> Maybe the confusion is that you're manipulating time while also travelling through it?

I guess this is where the confusion comes from. Without thinking about the internals, I just had expected `late` to behave another way and I wonder if it's somehow possible to get that behaviour, because it's more natural in that case

**Comment by yaxu** [comment]
Maybe this is it? https://slab.org/2020/02/21/time-spirals/

I think it would be possible to solve this at the same time as allowing different patterns to have different tempos, with some re-syncing mechanism. It seems to me that this is all about supporting a stateful approach to time.
